### DISCUSSION TITLE
How to configure https connection with path through environment variable?

### DISCUSSION THREAD

**Discussion by Kagemusha-san** [apache/airflow#17780] [Q&A] [answered]
I'm failing to achieve what seems to be a simple task: defining an https connection via environment variable with path like https://localhost:443/gbs.

We're using a `SimpleHttpOperator` to make a call to /api/datapump/load endpoint
```
SimpleHttpOperator(
   task_id='test_task',
   method='POST',
   http_conn_id='gbs_api',
   endpoint='/api/datapump/load',
   data=json.dumps( {
	"Id": 52,
	"EntityNames": ["pet"]
   }),
   dag=dag,
   do_xcom_push=True,
   headers = {
       'Authorization':'Bearer {{ ti.xcom_pull(task_ids="login_task") }}',
       'Content-Type': 'application/json'
   },
   response_check=lambda response: response.json()['Status'] == 200,
)
```
For http, gbs_api connection is defined in environment as `AIRFLOW_CONN_GBS_API=http://localhost:4048/` which works as expected.

For https, were we serve our services behind an nginx proxy, the base URL is https://localhost:443/gbs. Since Airflow's [connection URI format](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#connection-uri-format) expects the schema after the port, I'm not able to simply pass the URL as is.

First thing I tried was to URI-escape, as suggested [here](https://issues.apache.org/jira/browse/AIRFLOW-2910?focusedCommentId=16832468&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16832468), resulting in  `AIRFLOW_CONN_GBS_API='http://https%3a%2f%2flocalhost%3a443%2fgbs'` resp. `AIRFLOW_CONN_GBS_API='https%3a%2f%2flocalhost%3a443%2fgbs'`. Neither one works and yields following error
```
{http.py:111} INFO - Calling HTTP method
{http.py:140} INFO - Sending 'POST' to url: https://localhost:443/gbs':///api/datapump/load
{http.py:154} ERROR - HTTP error: Not Found
```

Maybe it has to be added to the host, I thought, and went for it: `AIRFLOW_CONN_GBS_API=http://localhost%2Fgbs:443/https`. But of course it then tries to connect to https://localhost/gbs:443/api/datapump/load which isn't quite right.

So I got a little creative and tried appending it to the schema, hoping that only the first /* part after the port would be recognized as schema `AIRFLOW_CONN_GBS_API='http://localhost:443/https/gbs` (also tried escaped variant with %2fgbs) but unfortunately, it recognizes everything after the port as schema and tries to connect to https/gbs://localhost:443. 

I also recognized the `extra` field in connection configuration but according to [http operator docs](https://airflow.apache.org/docs/apache-airflow-providers-http/stable/connections/http.html) it's used for headers. Of course I also read [the docs regarding connections](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html) but I haven't found anything in that regard, or maybe I overlooked it.

Long story short: how can I define a connection like https://locahost:443/gbs via environment variable?

**Comment by potiuk** [comment]
Yeah. For historical reasons it is far more complex than you would think.

`AIRFLOW_CONN_GBS_API='http://https%3a%2f%2flocalhost:443'`

Should work.

UPDATE: You still need to use `endpoint=gbs` in this case - see below

**Comment by potiuk** [comment, marked as answer]
You also need to use ``endpoint`` in your  task execution - HTTP Hook contains `endpoint` parameter so you should pass `endpoint=gbs` in your task, and you can specify ``http://localhost:443/https` as your connection URI ( and yes, I know it's how it should be, we will likely change it in the future)

UPDATE: Seems that endpoint is always needed if you want to use SimpleHTTPOperator

**Comment by potiuk** [comment]
I created a PR to clarify that https://github.com/apache/airflow/pull/17783 in the future version of the operator's `HowTo` guide.

**Comment by Kagemusha-san** [comment]
Using http://localhost:443/https as environment variable and prepending gbs to the `endpoint` parameter of `SimpleHttpOperator` works. Since we're also generating DAGs in our application we should be able to distinguish between http and https and set the `endpoint `parameter accordingly.

**Reply by potiuk** [reply]
Yep. The (already merged) #17783 already reflects that :) - thanks for double-checking!

**Comment by MM-Lehmann** [comment]
wouldn't it be much easier if we had a dedicated HTTPS connector/provider? I am stumbling over this many a times in all sorts of use cases. The whole world is using https now...

**Reply by potiuk** [reply]
It has already been addressed with https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#json-format-example - in Airflow 2.3 where you can specify the schema to use in JSON form. @MM-Lehmann - would you be so kind to verify if the solution is good for you and confirm it here? You seem to think that what we have is not enough. I'd appreciate your thoughts after trying this one.

**Reply by MM-Lehmann** [reply]
We haven't migrated to 2.3.x just yet because sqlalchemy-exasol isn't compatible anymore (might have to switch to pyexasol, which isn't trivial). I will test as soon as 2.3.x is an option for us.
Thanks for the input, though.
Do I understand correctly, that I can just put `https` down as `schema` and the URI will be correct? What about SSL handling?

**Reply by potiuk** [reply]
We are using whatever requests library uses under the hood
